Look for problems in the system

mnowaksuse opened the pull request:
POO#11930

This new sub problem_detection() detects problems in the installed
system and uploads log to worker. Currently it looks into systemd
journal, coredumpctl, systemctl et all for core dumps, signs of python
traceback, general problems in the journal etc. (Other proposals
welcome.)

I implemented POO#11930 not as an enhancement to textinfo.pm test but as
a function which might be called from different places (installed system
and even installer itself), and multiple times. Atm I execute it from
`console/consoletest_setup.pm` and `x11/shutdown.pm`. Problem of current
approach is that it adds 5-6 minutes to the test suite per execution.

Verification run: http://assam.suse.cz/tests/2678#downloads.

okurz commented:
Interesting approach. I like the idea in general. What I am not so sure about is that you are calling this in "consoletest_setup" which is a part of a lot of scenarios and will further slow down the execution in cases when everything is just fine. Also, as complexity increases in your problem analysis, jobs can also fail in there. I suggest to better separate what is "setup", what is "gather generic information" and "do analysis of a real problem".

For a start, maybe you can even in the current state move out of the logs gathering from "consoletest_setup" and make that module just what it says, a "setup", e.g. only including setting up the serial port, consoles, making sure packagekit is stopped, etc.

The sub `problem_detection` by itself looks good but can we start by calling it in post_fail_hook for a start and discuss later where it should be called even in case of passing test modules?

mnowaksuse commented:
http://assam.suse.cz/tests/2774#step/consoletest_setup/31

Removed some duplicities from `console/consoletest_setup`, moved analysis to `export_logs()`. `problem_detection()` is now being called from `export_logs()` only but can be called from tests as well where needed. `problem_detection()`'s analysis is now bit faster as it's result is packed and uploaded only once.

Still:
1. Calling `save_and_upload_log()` feels overcomplicated.
2. My feeling is that once core dump is found on the system, we reviewer should be notified as it's probably something we want to be fixed anyway. But using `assert_script_run()` just end the sub (as it should): http://assam.suse.cz/tests/2777#step/consoletest_setup/57.

okurz commented on lib/opensusebasetest.pm:
```diff
@@ -25,21 +25,98 @@ sub post_run_hook {
+    }
+    else {
+        script_run("$cmd | tee $file");
+    }
```
perl beginner question: As `$args` should always exist as a variable, does it work to just write

```perl
script_run("$cmd | tee $file", $args->{timeout});
```
in any case without getting a warning?

mnowaksuse replied:
It actually works as `use warnings` is not present. Thanks for the hint.

okurz commented on tests/console/consoletest_setup.pm:
```diff
@@ -66,28 +66,18 @@ sub run() {
     type_string "chown $username /dev/$serialdev\n";
-    # Export the existing status of running tasks and system load for future reference (fail would export it again)
-    script_run "ps axf > /tmp/psaxf.log";
-    script_run "cat /proc/loadavg > /tmp/loadavg_consoletest_setup.txt";
```
you don't call these anywhere now? I think they can be helpful sometimes so should be preserved? Or am I missing it?

mnowaksuse replied:
They were duplicated in `export_logs()` I removed them from the test and left in the sub. Would you prefer to keep them duplicated?

nilxam replied:
please keep them here, we want to see them anyway not only when the test failed.

okurz replied:
yes, it's good to have them so that we can compare these logs against the ones of a failing module.

okurz commented:
can't access your machine, is it down?

mnowaksuse commented:
> can't access your machine, is it down?

It's back online now.

mnowaksuse commented:
New verify run: http://assam.suse.cz/tests/2785#step/consoletest_setup/32.

nilxam commented on tests/console/consoletest_setup.pm:
```diff
@@ -70,16 +70,14 @@ sub run() {
     script_run "ps axf > /tmp/psaxf.log";
     script_run "cat /proc/loadavg > /tmp/loadavg_consoletest_setup.txt";
 
-    # Just after the setup: let's see the network configuration
```
we also need it in each succeeded run, please do not remove it or only uploading when the test failed.

okurz commented:
LGTM

wanna remove the WIP label?

nilxam commented:
LGTM
